src/weights.jl

```diff
@@ -2,44 +2,189 @@
+
+Frequency weights describe the number of cases (or frequency) in which each observation
+was observed. These weight may also be referred to as case weights or repeat weights.
+"""
+FrequencyWeights{S<:Real, V<:RealVector}(vs::V, s::S=sum(vs)) =
+    FrequencyWeights{S, eltype(vs), V}(vs, s)
+
+"""
+    fweights(vs)
+
+Construct a `FrequencyWeights` vector from a given array.
+See the documentation for `FrequencyWeights` for more details.
+"""
+fweights(vs::RealVector) = FrequencyWeights(vs)
+fweights(vs::RealArray) = FrequencyWeights(vec(vs))
+
+"""
+    cfactor(w::FrequencyWeights, true)
+
+``\\frac{1}{\sum{w} - 1}``
 """
-    weights(vs)
+cfactor(w::FrequencyWeights, ::Type{Val{true}}) = 1 / (sum(w) - 1)
 
-Construct a `WeightVec` from a given array.
+@weights ProbabilityWeights
+
+"""
+    ProbabilityWeights(vs, wsum=sum(vs))
+
+Construct a `ProbabilityWeights` vector with weight values `vs` and sum of weights `wsum`.
+
+Probability weights represent the inverse of the sampling probability for each observation,
+providing a correction mechanism for under- or over-sampling certain population groups.
+These weight may also be referred to as sampling weights.
+"""
+ProbabilityWeights{S<:Real, V<:RealVector}(vs::V, s::S=sum(vs)) =
+    ProbabilityWeights{S, eltype(vs), V}(vs, s)
+
+"""
+    pweights(vs)
+
+Construct a `ProbabilityWeights` vector from a given array.
+See the documentation for `ProbabilityWeights` for more details.
 """
-weights(vs::RealVector) = WeightVec(vs)
-weights(vs::RealArray) = WeightVec(vec(vs))
+pweights(vs::RealVector) = ProbabilityWeights(vs)
+pweights(vs::RealArray) = ProbabilityWeights(vec(vs))
 
-eltype(wv::WeightVec) = eltype(wv.values)
-length(wv::WeightVec) = length(wv.values)
-values(wv::WeightVec) = wv.values
-sum(wv::WeightVec) = wv.sum
-isempty(wv::WeightVec) = isempty(wv.values)
+"""
+    cfactor(w::ProbabilityWeights, true)
+
+``\\frac{n}{(n - 1) \sum w}`` where `n = length(w)`
+"""
+function cfactor(w::ProbabilityWeights, ::Type{Val{true}})
+    s = sum(w)
+    n = length(w)
+    return n / (s * (n - 1))
+end
 
-Base.getindex(wv::WeightVec, i) = getindex(wv.values, i)
-Base.size(wv::WeightVec) = size(wv.values)
+"""
+    eweights(n, [λ])
+
+Constructs an `AnalyticWeights` vector with a desired length `n` and smoothing factor `λ`,
+where each element is set to ``λ * (1 - λ)^(1 - i)``.
 
+# Arguments
+* `n::Integer`: the desired length of the `Weights`
+* `λ::Real`: a smoothing factor or rate parameter between 0 and 1.
+    As this value approaches 0 the resulting weights will be almost equal,
+    while values closer to 1 will put higher weight on the end elements of the vector.
+"""
+function eweights(n::Integer, λ::Real=0.99)
+    n > 0 || throw(ArgumentError("cannot construct weights of length < 1"))
+    0 <= λ <= 1 || throw(ArgumentError("smoothing factor must be between 0 and 1"))
+    w0 = map(i -> λ * (1 - λ)^(1 - i), 1:n)
+    aweights(w0)
```

Comment:
Are these really analytical weights? The docs should mention this if that's the case. Can you explain what exponential weights are used for?